In [1]:
from telco_setup import telco_imports
telco_imports()


import os
import seaborn as sns
import matplotlib.pyplot as plt
import pandas as pd
import numpy as np
import math as m
from scipy import stats
from stat_tools import *
from sklearn.model_selection import train_test_split
import acquire
import prepare
from sklearn.metrics import accuracy_score, precision_score, recall_score, classification_report
import lux
    


In [2]:
import os
import seaborn as sns
import matplotlib.pyplot as plt
import pandas as pd
import numpy as np
import math as m
from scipy import stats
from stat_tools import *
from sklearn.model_selection import train_test_split
import acquire
import prepare
from sklearn.metrics import accuracy_score, precision_score, recall_score, classification_report
# grab Decision Tree Classifier and some helping friends from sklearn.tree
from sklearn.tree import DecisionTreeClassifier, plot_tree, export_text
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import confusion_matrix
from sklearn.neighbors import KNeighborsClassifier
from sklearn.linear_model import LogisticRegression


In [3]:
df = prepare.fix_nulls()

/Users/davidhowell/codeup-data-scinence/telco-churn/prepare.py:113: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['total_charges'][df['null_charges'] == True] = df['monthly_charges'][df['null_charges'] == True]


In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7043 entries, 0 to 7042
Data columns (total 22 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   customer_id            7043 non-null   object 
 1   gender                 7043 non-null   object 
 2   senior_citizen         7043 non-null   int64  
 3   partner                7043 non-null   object 
 4   dependents             7043 non-null   object 
 5   tenure                 7043 non-null   int64  
 6   phone_service          7043 non-null   object 
 7   multiple_lines         7043 non-null   object 
 8   internet_service_type  7043 non-null   object 
 9   online_security        7043 non-null   object 
 10  online_backup          7043 non-null   object 
 11  device_protection      7043 non-null   object 
 12  tech_support           7043 non-null   object 
 13  streaming_tv           7043 non-null   object 
 14  streaming_movies       7043 non-null   object 
 15  cont

In [5]:
df = prepare.prep_telco_1(df)
df

senior  partner  dependents  tenure  e_bill  monthly_charges  \
0          0        1           1       9       1            65.60   
1          0        0           0       9       0            59.90   
2          0        0           0       4       1            73.90   
3          1        1           0      13       1            98.00   
4          1        1           0       3       1            83.90   
...      ...      ...         ...     ...     ...              ...   
7038       0        0           0      13       0            55.15   
7039       0        1           0      22       1            85.10   
7040       0        0           0       2       1            50.30   
7041       0        1           1      67       0            67.85   
7042       0        1           1      63       0            59.00   

      total_charges  churn1  DSL  Fiber  one_year  two_year  bank_transfer  \
0            593.30       0    1      0         1         0              0   
1            542.40       0    1      0         0         0              0   
2            280.85       1    0      1         0         0              0   
3           1237.85       1    0      1         0         0              0   
4            267.40       1    0      1         0         0              0   
...             ...     ...  ...    ...       ...       ...            ...   
7038         742.90       0    1      0         1         0              0   
7039        1873.70       1    0      1         0         0              0   
7040          92.75       0    1      0         0         0              0   
7041        4627.65       0    1      0         0         1              0   
7042        3707.60       0    1      0         0         1              0   

      cc  e_check  add_ons  
0      0        0        2  
1      0        0        0  
2      0        1        1  
3      0        1        2  
4      0        0        1  
...   ..      ...      ...  
7038   0        0        2  
7039   0        1        0  
7040   0        0        1  
7041   0        0        3  
7042   0        1        3  

[7043 rows x 16 columns]

In [6]:
train, validate, test = prepare.split_data(df, 'churn1')

Prepared df: (7043, 16)

Train: (4057, 16)
Validate: (1409, 16)
Test: (1409, 16)


In [7]:
train

senior  partner  dependents  tenure  e_bill  monthly_charges  \
5911       0        0           1      26       1            39.95   
2479       0        1           0      47       0            26.90   
5889       0        0           0       1       1            50.45   
6087       0        0           0      69       1            60.05   
785        0        0           0      51       0            87.55   
...      ...      ...         ...     ...     ...              ...   
5269       1        1           1      60       1            80.95   
6102       0        1           0       2       0            79.70   
356        1        0           0      52       1            50.50   
1756       0        0           0       4       1            70.20   
2127       0        0           0      35       0            55.25   

      total_charges  churn1  DSL  Fiber  one_year  two_year  bank_transfer  \
5911        1023.75       0    1      0         1         0              1   
2479        1250.85       0    0      0         1         0              0   
5889          50.45       1    1      0         0         0              0   
6087        4176.70       0    1      0         0         1              1   
785         4475.90       0    0      1         1         0              0   
...             ...     ...  ...    ...       ...       ...            ...   
5269        4859.10       0    0      1         0         0              0   
6102         165.00       1    0      1         0         0              1   
356         2566.30       0    1      0         0         0              0   
1756         237.95       1    0      1         0         0              0   
2127        1924.10       0    1      0         0         1              0   

      cc  e_check  add_ons  
5911   0        0        1  
2479   0        1        0  
5889   0        0        1  
6087   0        0        3  
785    1        0        0  
...   ..      ...      ...  
5269   0        1        1  
6102   0        0        2  
356    0        1        3  
1756   0        1        0  
2127   0        0        1  

[4057 rows x 16 columns]

In [8]:
X_cols = train.columns.to_list()
X_cols.remove('churn1')
y_col = 'churn1'

In [9]:
X_cols

['senior',
 'partner',
 'dependents',
 'tenure',
 'e_bill',
 'monthly_charges',
 'total_charges',
 'DSL',
 'Fiber',
 'one_year',
 'two_year',
 'bank_transfer',
 'cc',
 'e_check',
 'add_ons']

In [10]:
y_col

'churn1'

In [11]:
X_train, y_train = train[X_cols], train[y_col]
X_validate, y_validate = validate[X_cols], validate[y_col]
X_test, y_test = test[X_cols], test[y_col]

In [12]:
# make the thing
dt_model_1 = DecisionTreeClassifier(random_state=123)
# fit the thing
dt_model_1.fit(X_train, y_train)
# use the thing
print(f'training score: {dt_model_1.score(X_train, y_train):.2%}')
print(f'validate score: {dt_model_1.score(X_validate, y_validate):.2%}')

training score: 99.70%
validate score: 72.25%


In [13]:
# make the thing
dt_model_2 = DecisionTreeClassifier(random_state=123, max_depth=5)
# fit the thing
dt_model_2.fit(X_train, y_train)
# use the thing
print(f'training score: {dt_model_2.score(X_train, y_train):.2%}')
print(f'validate score: {dt_model_2.score(X_validate, y_validate):.2%}')

training score: 79.66%
validate score: 78.50%


In [14]:
# make the thing
dt_model_3 = DecisionTreeClassifier(random_state=123, max_depth=3)
# fit the thing
dt_model_3.fit(X_train, y_train)
# use the thing
print(f'training score: {dt_model_3.score(X_train, y_train):.2%}')
print(f'validate score: {dt_model_3.score(X_validate, y_validate):.2%}')

training score: 78.06%
validate score: 79.28%


In [15]:
# rf_model_1 = RandomForestClassifier(random_state=123,)

In [16]:
# from modeling import *

In [17]:
int(y_train.mode())

0

In [33]:
def random_forest_models(num_models, rand_st=123, positive=1, max_samp=1.0, trees=100):
    '''
    random_forest_models is a function that:
        
        Takes in:   num_models=  >> The number of rf models 
                                  you want to create  ;dtype(int)
                       rand_st=  >> Random State  
                                  ;dtype(int) = 123 unless specified
                      positive=  >> what is the positive test 
                                  (0 or 1)
                      max_samp=  >> maximum samples per tree
                                  ;dtype(int, float) = (default)1.0
                                  if int: = number of samples
                                  if float: = percentage of total samples
                         trees=  >> n_estimators: number of trees in the forest
        
Assumed variables apply:
    
                  train: training dataset
               validate: validate dataset
                   test: test dataset

                 X_cols = df.columns.drop('target_y').to_list()
                  y_col = 'target_y'

                X_train = train[X_cols]
                y_train = train[y_col]
             X_validate = validate[X_cols]
             y_validate = validate[y_col]
                 X_test = test[X_cols]
                 y_test = test[y_col]
                 
        Returns: a DataFrame with predictions for each model
    '''
    b = int(y_train.mode())
    preds = pd.DataFrame({
    'actual': y_train,
    'baseline': b,
    })
    depth = 11 #num_models * 2 + 1
#     fig, ax = plt.subplot(nrows = num_models,n)
    best = 1
    for i in range(1, num_models+1):
        depth -= 1
        name = f'model_{i}_depth_{depth}'
        
        rf = RandomForestClassifier(random_state = rand_st, 
                                    min_samples_leaf = i, 
                                    max_depth = depth,
                                    max_samples = max_samp,
                                    n_estimators = trees
                                   )
        rf.fit(X_train, y_train)
        
        preds[name] = rf.predict(X_train)
#         val_name = f'{name}_validate'
        TN, FP, FN, TP = confusion_matrix(preds.actual, preds[name]).ravel()
        print(f'\n{name}\n\n {rf}')
        confusion(TN=TN, TP=TP, FN=FN, FP=FP)
        print(f'Validation score is: {rf.score(X_validate, y_validate):.2%}')
        print('______________________________')
#         preds[val_name] = rf.predict(X_val)
#         plt.subplot(i,i,12)
#         plt.title(f'{name} feature importances')
#         plt.barh(X_train.columns, rf.feature_importances_)
#         plt.show
                
    return preds

In [20]:
def confusion(TN, TP, FN, FP):
    acc = (TP+TN)/(TP+TN+FP+FN)
    pre = (TP/(TP+FP))
    NPV = (TN/(TN+FN))
    rec = (TP/(TP+FN))
    spe = (TN/(TN+FP))
    f1s = stats.hmean([(TP/(TP+FP)),(TP/(TP+FN))])
    print(
    f'''
    _______________________________________________________________________________________
    
    True Positive = {TP} ---- False Positive = {FP}
    True Negative = {TN} ---- False Negative = {FN}
    
    Out of {TP+FN+FP+TN} predictions -- Correct predictions = {TP+TN} (True Pos + True Neg) 
    
    REAL POSITIVE = (TP + FN) = {TP+FN} ---- PREDICTED POSITIVE = (TP + FP) = {TP+FP}
    
    REAL NEGATIVE = (TN + FP) = {TN+FP} ---- PREDICTED NEGATIVE = (TN + FN) = {TN+FN}
     
        Accuracy = {acc:.2%} -->> Correct Predictions / Total Predictions
       Precision = {pre:.2%} -->> True Positive / Predicted Positive
             NPV = {NPV:.2%} -->> True Negative / Predicted Negative
          Recall = {rec:.2%} -->> True Positive / Real Positive
     Specificity = {spe:.2%} -->> True Negative / Real Negative
        f1-score = {f1s:.2%} -->> Harmonic Mean of Precision and Recall
    _______________________________________________________________________________________
    '''
    )

In [34]:
preds = random_forest_models(9)


model_1_depth_10

 RandomForestClassifier(max_depth=10, max_samples=1.0, random_state=123)

    _______________________________________________________________________________________
    
    True Positive = 779 ---- False Positive = 167
    True Negative = 2776 ---- False Negative = 335
    
    Out of 4057 predictions -- Correct predictions = 3555 (True Pos + True Neg) 
    
    REAL POSITIVE = (TP + FN) = 1114 ---- PREDICTED POSITIVE = (TP + FP) = 946
    
    REAL NEGATIVE = (TN + FP) = 2943 ---- PREDICTED NEGATIVE = (TN + FN) = 3111
     
        Accuracy = 87.63% -->> Correct Predictions / Total Predictions
       Precision = 82.35% -->> True Positive / Predicted Positive
             NPV = 89.23% -->> True Negative / Predicted Negative
          Recall = 69.93% -->> True Positive / Real Positive
     Specificity = 94.33% -->> True Negative / Real Negative
        f1-score = 75.63% -->> Harmonic Mean of Precision and Recall
    __________________________________________________

In [29]:
preds

actual  baseline  model_1_depth_10  model_2_depth_9  model_3_depth_8  \
5911       0         0                 0                0                0   
2479       0         0                 0                0                0   
5889       1         0                 1                1                1   
6087       0         0                 0                0                0   
785        0         0                 0                0                0   
...      ...       ...               ...              ...              ...   
5269       0         0                 0                0                0   
6102       1         0                 1                1                1   
356        0         0                 0                0                0   
1756       1         0                 1                1                1   
2127       0         0                 0                0                0   

      model_4_depth_7  model_5_depth_6  model_6_depth_5  model_7_depth_4  \
5911                0                0                0                0   
2479                0                0                0                0   
5889                1                1                0                0   
6087                0                0                0                0   
785                 0                0                0                0   
...               ...              ...              ...              ...   
5269                0                0                0                0   
6102                1                1                1                1   
356                 0                0                0                0   
1756                1                1                1                1   
2127                0                0                0                0   

      model_8_depth_3  model_9_depth_2  
5911                0                0  
2479                0                0  
5889                0                0  
6087                0                0  
785                 0                0  
...               ...              ...  
5269                0                0  
6102                1                0  
356                 0                0  
1756                1                1  
2127                0                0  

[4057 rows x 11 columns]

In [30]:
for i in preds:
    cm = confusion_matrix(preds.actual,preds[i])
    print(f'''{i}, 
{cm}
_________________''')

actual, 
[[2943    0]
 [   0 1114]]
_________________
baseline, 
[[2943    0]
 [1114    0]]
_________________
model_1_depth_10, 
[[2776  167]
 [ 335  779]]
_________________
model_2_depth_9, 
[[2719  224]
 [ 421  693]]
_________________
model_3_depth_8, 
[[2693  250]
 [ 457  657]]
_________________
model_4_depth_7, 
[[2685  258]
 [ 499  615]]
_________________
model_5_depth_6, 
[[2693  250]
 [ 537  577]]
_________________
model_6_depth_5, 
[[2723  220]
 [ 619  495]]
_________________
model_7_depth_4, 
[[2752  191]
 [ 681  433]]
_________________
model_8_depth_3, 
[[2782  161]
 [ 740  374]]
_________________
model_9_depth_2, 
[[2906   37]
 [ 971  143]]
_________________


In [35]:
preds1 = random_forest_models(9, max_samp=.5)


model_1_depth_10

 RandomForestClassifier(max_depth=10, max_samples=0.5, random_state=123)

    _______________________________________________________________________________________
    
    True Positive = 726 ---- False Positive = 186
    True Negative = 2757 ---- False Negative = 388
    
    Out of 4057 predictions -- Correct predictions = 3483 (True Pos + True Neg) 
    
    REAL POSITIVE = (TP + FN) = 1114 ---- PREDICTED POSITIVE = (TP + FP) = 912
    
    REAL NEGATIVE = (TN + FP) = 2943 ---- PREDICTED NEGATIVE = (TN + FN) = 3145
     
        Accuracy = 85.85% -->> Correct Predictions / Total Predictions
       Precision = 79.61% -->> True Positive / Predicted Positive
             NPV = 87.66% -->> True Negative / Predicted Negative
          Recall = 65.17% -->> True Positive / Real Positive
     Specificity = 93.68% -->> True Negative / Real Negative
        f1-score = 71.67% -->> Harmonic Mean of Precision and Recall
    __________________________________________________


model_9_depth_2

 RandomForestClassifier(max_depth=2, max_samples=0.5, min_samples_leaf=9,
                       random_state=123)

    _______________________________________________________________________________________
    
    True Positive = 121 ---- False Positive = 29
    True Negative = 2914 ---- False Negative = 993
    
    Out of 4057 predictions -- Correct predictions = 3035 (True Pos + True Neg) 
    
    REAL POSITIVE = (TP + FN) = 1114 ---- PREDICTED POSITIVE = (TP + FP) = 150
    
    REAL NEGATIVE = (TN + FP) = 2943 ---- PREDICTED NEGATIVE = (TN + FN) = 3907
     
        Accuracy = 74.81% -->> Correct Predictions / Total Predictions
       Precision = 80.67% -->> True Positive / Predicted Positive
             NPV = 74.58% -->> True Negative / Predicted Negative
          Recall = 10.86% -->> True Positive / Real Positive
     Specificity = 99.01% -->> True Negative / Real Negative
        f1-score = 19.15% -->> Harmonic Mean of Precision and Recall
    __________

model_5_depth_6

 RandomForestClassifier(max_depth=6, max_samples=0.5, min_samples_leaf=5,
                       random_state=123)

    _______________________________________________________________________________________
    
    True Positive = 581 ---- False Positive = 254
    True Negative = 2689 ---- False Negative = 533
    
    Out of 4057 predictions -- Correct predictions = 3270 (True Pos + True Neg) 
    
    REAL POSITIVE = (TP + FN) = 1114 ---- PREDICTED POSITIVE = (TP + FP) = 835
    
    REAL NEGATIVE = (TN + FP) = 2943 ---- PREDICTED NEGATIVE = (TN + FN) = 3222
     
        Accuracy = 80.60% -->> Correct Predictions / Total Predictions
       Precision = 69.58% -->> True Positive / Predicted Positive
             NPV = 83.46% -->> True Negative / Predicted Negative
          Recall = 52.15% -->> True Positive / Real Positive
     Specificity = 91.37% -->> True Negative / Real Negative
        f1-score = 59.62% -->> Harmonic Mean of Precision and Recall
    _______________________________________________________________________________________
    
Validation score is: 79.99%
______________________________

In [53]:
knn1 = KNeighborsClassifier()
knn1.fit(X_train, y_train)
knn_pred_1 = knn.predict(X_train)
knn_pred_1

array([0, 0, 0, ..., 0, 1, 0])

In [54]:
b = int(y_train.mode())
knn_preds = pd.DataFrame({
    'actual': y_train,
    'baseline': b,
    })

In [55]:
knn_preds['knn_pred_1'] = knn_pred_1

In [56]:
knn_preds

actual  baseline  knn_pred_1
5911       0         0           0
2479       0         0           0
5889       1         0           0
6087       0         0           0
785        0         0           0
...      ...       ...         ...
5269       0         0           0
6102       1         0           1
356        0         0           0
1756       1         0           1
2127       0         0           0

[4057 rows x 3 columns]

In [57]:
pd.crosstab(knn_preds.actual, knn_preds.knn_pred_1)

knn_pred_1     0    1
actual               
0           2942    1
1            610  504

In [61]:
knn2 = KNeighborsClassifier(n_neighbors=2)
knn2.fit(X_train, y_train)
knn_pred_2 = knn.predict(X_train)
# knn_pred_2

knn_preds['knn_pred_2'] = knn_pred_2

pd.crosstab(knn_preds.actual, knn_preds.knn_pred_2)

knn_pred_2     0    1
actual               
0           2942    1
1            610  504

In [62]:
knn3 = KNeighborsClassifier(n_neighbors=3)
knn3.fit(X_train, y_train)
knn_pred_3 = knn.predict(X_train)
# knn_pred_2

knn_preds['knn_pred_3'] = knn_pred_3

pd.crosstab(knn_preds.actual, knn_preds.knn_pred_3)

knn_pred_3     0    1
actual               
0           2942    1
1            610  504

In [84]:
knn1 = KNeighborsClassifier()
knn1.fit(X_train, y_train)
knn_pred_1 = knn.predict(X_train)
# knn_pred_2

knn_preds['knn_pred_1'] = knn_pred_1

pd.crosstab(knn_preds.actual, knn_preds.knn_pred_1)

knn_pred_1     0    1
actual               
0           2942    1
1            610  504

In [66]:
knn_preds

actual  baseline  knn_pred_1  knn_pred_2  knn_pred_3  knn_pred_5  \
5911       0         0           0           0           0           0   
2479       0         0           0           0           0           0   
5889       1         0           0           0           0           0   
6087       0         0           0           0           0           0   
785        0         0           0           0           0           0   
...      ...       ...         ...         ...         ...         ...   
5269       0         0           0           0           0           0   
6102       1         0           1           1           1           1   
356        0         0           0           0           0           0   
1756       1         0           1           1           1           1   
2127       0         0           0           0           0           0   

      knn_pred_10  
5911            0  
2479            0  
5889            0  
6087            0  
785             0  
...           ...  
5269            0  
6102            1  
356             0  
1756            1  
2127            0  

[4057 rows x 7 columns]

In [67]:
from sklearn.model_selection import GridSearchCV

In [85]:
knn1.get_params()

{'algorithm': 'auto',
 'leaf_size': 30,
 'metric': 'minkowski',
 'metric_params': None,
 'n_jobs': None,
 'n_neighbors': 5,
 'p': 2,
 'weights': 'uniform'}

In [86]:
gridsearch = GridSearchCV(knn1, param_grid={'leaf_size': np.arange(1,30).tolist(), 'n_nieghbors': [1,2,3,4,5,6,7,8,9,10]})

In [87]:
gridsearch.fit(X_train, y_train)

ValueError: Invalid parameter n_nieghbors for estimator KNeighborsClassifier(leaf_size=1). Check the list of available parameters with `estimator.get_params().keys()`.

In [83]:
gridsearch.get_params()

{'cv': None,
 'error_score': nan,
 'estimator__algorithm': 'auto',
 'estimator__leaf_size': 30,
 'estimator__metric': 'minkowski',
 'estimator__metric_params': None,
 'estimator__n_jobs': None,
 'estimator__n_neighbors': 1,
 'estimator__p': 2,
 'estimator__weights': 'uniform',
 'estimator': KNeighborsClassifier(n_neighbors=1),
 'n_jobs': None,
 'param_grid': {'leaf_size': [1,
   2,
   3,
   4,
   5,
   6,
   7,
   8,
   9,
   10,
   11,
   12,
   13,
   14,
   15,
   16,
   17,
   18,
   19,
   20,
   21,
   22,
   23,
   24,
   25,
   26,
   27,
   28,
   29],
  'n_nieghbors': [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]},
 'pre_dispatch': '2*n_jobs',
 'refit': True,
 'return_train_score': False,
 'scoring': None,
 'verbose': 0}

In [92]:
X_cols = ['senior',
 'partner',
 'dependents',
 'e_bill',
 'tenure',
 'DSL',
 'Fiber',
 'bank_transfer',
 'cc',
 'e_check',
 'add_ons']

In [93]:
X_train, y_train = train[X_cols], train[y_col]
X_validate, y_validate = validate[X_cols], validate[y_col]
X_test, y_test = test[X_cols], test[y_col]

In [94]:
preds2 = random_forest_models(9, trees=1000)


model_1_depth_10

 RandomForestClassifier(max_depth=10, max_samples=1.0, n_estimators=1000,
                       random_state=123)

    _______________________________________________________________________________________
    
    True Positive = 743 ---- False Positive = 176
    True Negative = 2767 ---- False Negative = 371
    
    Out of 4057 predictions -- Correct predictions = 3510 (True Pos + True Neg) 
    
    REAL POSITIVE = (TP + FN) = 1114 ---- PREDICTED POSITIVE = (TP + FP) = 919
    
    REAL NEGATIVE = (TN + FP) = 2943 ---- PREDICTED NEGATIVE = (TN + FN) = 3138
     
        Accuracy = 86.52% -->> Correct Predictions / Total Predictions
       Precision = 80.85% -->> True Positive / Predicted Positive
             NPV = 88.18% -->> True Negative / Predicted Negative
          Recall = 66.70% -->> True Positive / Real Positive
     Specificity = 94.02% -->> True Negative / Real Negative
        f1-score = 73.09% -->> Harmonic Mean of Precision and Recall
    ________


model_9_depth_2

 RandomForestClassifier(max_depth=2, max_samples=1.0, min_samples_leaf=9,
                       n_estimators=1000, random_state=123)

    _______________________________________________________________________________________
    
    True Positive = 212 ---- False Positive = 70
    True Negative = 2873 ---- False Negative = 902
    
    Out of 4057 predictions -- Correct predictions = 3085 (True Pos + True Neg) 
    
    REAL POSITIVE = (TP + FN) = 1114 ---- PREDICTED POSITIVE = (TP + FP) = 282
    
    REAL NEGATIVE = (TN + FP) = 2943 ---- PREDICTED NEGATIVE = (TN + FN) = 3775
     
        Accuracy = 76.04% -->> Correct Predictions / Total Predictions
       Precision = 75.18% -->> True Positive / Predicted Positive
             NPV = 76.11% -->> True Negative / Predicted Negative
          Recall = 19.03% -->> True Positive / Real Positive
     Specificity = 97.62% -->> True Negative / Real Negative
        f1-score = 30.37% -->> Harmonic Mean of Precision and Re

In [95]:
preds3 = random_forest_models(9, trees=1000, max_samp=.6)


model_1_depth_10

 RandomForestClassifier(max_depth=10, max_samples=0.6, n_estimators=1000,
                       random_state=123)

    _______________________________________________________________________________________
    
    True Positive = 712 ---- False Positive = 193
    True Negative = 2750 ---- False Negative = 402
    
    Out of 4057 predictions -- Correct predictions = 3462 (True Pos + True Neg) 
    
    REAL POSITIVE = (TP + FN) = 1114 ---- PREDICTED POSITIVE = (TP + FP) = 905
    
    REAL NEGATIVE = (TN + FP) = 2943 ---- PREDICTED NEGATIVE = (TN + FN) = 3152
     
        Accuracy = 85.33% -->> Correct Predictions / Total Predictions
       Precision = 78.67% -->> True Positive / Predicted Positive
             NPV = 87.25% -->> True Negative / Predicted Negative
          Recall = 63.91% -->> True Positive / Real Positive
     Specificity = 93.44% -->> True Negative / Real Negative
        f1-score = 70.53% -->> Harmonic Mean of Precision and Recall
    ________


model_9_depth_2

 RandomForestClassifier(max_depth=2, max_samples=0.6, min_samples_leaf=9,
                       n_estimators=1000, random_state=123)

    _______________________________________________________________________________________
    
    True Positive = 212 ---- False Positive = 69
    True Negative = 2874 ---- False Negative = 902
    
    Out of 4057 predictions -- Correct predictions = 3086 (True Pos + True Neg) 
    
    REAL POSITIVE = (TP + FN) = 1114 ---- PREDICTED POSITIVE = (TP + FP) = 281
    
    REAL NEGATIVE = (TN + FP) = 2943 ---- PREDICTED NEGATIVE = (TN + FN) = 3776
     
        Accuracy = 76.07% -->> Correct Predictions / Total Predictions
       Precision = 75.44% -->> True Positive / Predicted Positive
             NPV = 76.11% -->> True Negative / Predicted Negative
          Recall = 19.03% -->> True Positive / Real Positive
     Specificity = 97.66% -->> True Negative / Real Negative
        f1-score = 30.39% -->> Harmonic Mean of Precision and Re

In [96]:
preds4 = random_forest_models(9, trees=1000, max_samp=.3)


model_1_depth_10

 RandomForestClassifier(max_depth=10, max_samples=0.3, n_estimators=1000,
                       random_state=123)

    _______________________________________________________________________________________
    
    True Positive = 667 ---- False Positive = 213
    True Negative = 2730 ---- False Negative = 447
    
    Out of 4057 predictions -- Correct predictions = 3397 (True Pos + True Neg) 
    
    REAL POSITIVE = (TP + FN) = 1114 ---- PREDICTED POSITIVE = (TP + FP) = 880
    
    REAL NEGATIVE = (TN + FP) = 2943 ---- PREDICTED NEGATIVE = (TN + FN) = 3177
     
        Accuracy = 83.73% -->> Correct Predictions / Total Predictions
       Precision = 75.80% -->> True Positive / Predicted Positive
             NPV = 85.93% -->> True Negative / Predicted Negative
          Recall = 59.87% -->> True Positive / Real Positive
     Specificity = 92.76% -->> True Negative / Real Negative
        f1-score = 66.90% -->> Harmonic Mean of Precision and Recall
    ________


model_9_depth_2

 RandomForestClassifier(max_depth=2, max_samples=0.3, min_samples_leaf=9,
                       n_estimators=1000, random_state=123)

    _______________________________________________________________________________________
    
    True Positive = 204 ---- False Positive = 61
    True Negative = 2882 ---- False Negative = 910
    
    Out of 4057 predictions -- Correct predictions = 3086 (True Pos + True Neg) 
    
    REAL POSITIVE = (TP + FN) = 1114 ---- PREDICTED POSITIVE = (TP + FP) = 265
    
    REAL NEGATIVE = (TN + FP) = 2943 ---- PREDICTED NEGATIVE = (TN + FN) = 3792
     
        Accuracy = 76.07% -->> Correct Predictions / Total Predictions
       Precision = 76.98% -->> True Positive / Predicted Positive
             NPV = 76.00% -->> True Negative / Predicted Negative
          Recall = 18.31% -->> True Positive / Real Positive
     Specificity = 97.93% -->> True Negative / Real Negative
        f1-score = 29.59% -->> Harmonic Mean of Precision and Re

In [97]:
rfb = RandomForestClassifier(max_depth=15, max_samples=0.3, n_estimators=1000,
                       random_state=123)

In [98]:
rfb.fit(X_train, y_train)

RandomForestClassifier(max_depth=15, max_samples=0.3, n_estimators=1000,
                       random_state=123)

In [99]:
rfb_pred = rfb.predict(X_train)

In [110]:
# rfb.predict(X_train)

In [100]:
rfb_pred

array([0, 0, 1, ..., 0, 1, 0])

In [102]:
TN, FP, FN, TP = confusion_matrix(preds.actual, rfb_pred).ravel()
#         print(f'\n{name}\n\n {rf}')
confusion(TN=TN, TP=TP, FN=FN, FP=FP)
print(f'Validation score is: {rfb.score(X_validate, y_validate):.2%}')
print('______________________________')


    _______________________________________________________________________________________
    
    True Positive = 735 ---- False Positive = 150
    True Negative = 2793 ---- False Negative = 379
    
    Out of 4057 predictions -- Correct predictions = 3528 (True Pos + True Neg) 
    
    REAL POSITIVE = (TP + FN) = 1114 ---- PREDICTED POSITIVE = (TP + FP) = 885
    
    REAL NEGATIVE = (TN + FP) = 2943 ---- PREDICTED NEGATIVE = (TN + FN) = 3172
     
        Accuracy = 86.96% -->> Correct Predictions / Total Predictions
       Precision = 83.05% -->> True Positive / Predicted Positive
             NPV = 88.05% -->> True Negative / Predicted Negative
          Recall = 65.98% -->> True Positive / Real Positive
     Specificity = 94.90% -->> True Negative / Real Negative
        f1-score = 73.54% -->> Harmonic Mean of Precision and Recall
    _______________________________________________________________________________________
    
Validation score is: 78.28%
_____________________

In [103]:
rfb.feature_importances_

array([0.03525284, 0.04198151, 0.0349527 , 0.04655908, 0.52354688,
       0.02314095, 0.07471486, 0.02244981, 0.02291729, 0.05752419,
       0.11695989])

In [ ]:
# X_cols = ['senior',
#  'partner',
#  'dependents',
#  'e_bill',
#  'tenure',
#  'DSL',
#  'Fiber',
#  'bank_transfer',
#  'cc',
#  'e_check',
#  'add_ons']

In [104]:
X_cols = ['e_bill',
 'tenure',
 'Fiber',
 'bank_transfer',
 'e_check',
 'add_ons']

In [105]:
X_train, y_train = train[X_cols], train[y_col]
X_validate, y_validate = validate[X_cols], validate[y_col]
X_test, y_test = test[X_cols], test[y_col]

In [108]:
preds5 = random_forest_models(9, trees=1000, max_samp=.85)


model_1_depth_10

 RandomForestClassifier(max_depth=10, max_samples=0.85, n_estimators=1000,
                       random_state=123)

    _______________________________________________________________________________________
    
    True Positive = 720 ---- False Positive = 214
    True Negative = 2729 ---- False Negative = 394
    
    Out of 4057 predictions -- Correct predictions = 3449 (True Pos + True Neg) 
    
    REAL POSITIVE = (TP + FN) = 1114 ---- PREDICTED POSITIVE = (TP + FP) = 934
    
    REAL NEGATIVE = (TN + FP) = 2943 ---- PREDICTED NEGATIVE = (TN + FN) = 3123
     
        Accuracy = 85.01% -->> Correct Predictions / Total Predictions
       Precision = 77.09% -->> True Positive / Predicted Positive
             NPV = 87.38% -->> True Negative / Predicted Negative
          Recall = 64.63% -->> True Positive / Real Positive
     Specificity = 92.73% -->> True Negative / Real Negative
        f1-score = 70.31% -->> Harmonic Mean of Precision and Recall
    _______


model_9_depth_2

 RandomForestClassifier(max_depth=2, max_samples=0.85, min_samples_leaf=9,
                       n_estimators=1000, random_state=123)

    _______________________________________________________________________________________
    
    True Positive = 280 ---- False Positive = 108
    True Negative = 2835 ---- False Negative = 834
    
    Out of 4057 predictions -- Correct predictions = 3115 (True Pos + True Neg) 
    
    REAL POSITIVE = (TP + FN) = 1114 ---- PREDICTED POSITIVE = (TP + FP) = 388
    
    REAL NEGATIVE = (TN + FP) = 2943 ---- PREDICTED NEGATIVE = (TN + FN) = 3669
     
        Accuracy = 76.78% -->> Correct Predictions / Total Predictions
       Precision = 72.16% -->> True Positive / Predicted Positive
             NPV = 77.27% -->> True Negative / Predicted Negative
          Recall = 25.13% -->> True Positive / Real Positive
     Specificity = 96.33% -->> True Negative / Real Negative
        f1-score = 37.28% -->> Harmonic Mean of Precision and 

In [111]:
rf_try = RandomForestClassifier(max_depth=9, max_samples=0.85, min_samples_leaf=2,
                       n_estimators=1000, random_state=123)

In [112]:
rf_try.fit(X_train,y_train)

RandomForestClassifier(max_depth=9, max_samples=0.85, min_samples_leaf=2,
                       n_estimators=1000, random_state=123)

In [113]:
rf_try.predict(X_train)

array([0, 0, 1, ..., 0, 1, 0])

In [115]:
rf_try_val = rf_try.predict(X_validate)

In [118]:
y_validate.info()

<class 'pandas.core.series.Series'>
Int64Index: 1409 entries, 3554 to 3108
Series name: churn1
Non-Null Count  Dtype
--------------  -----
1409 non-null   int64
dtypes: int64(1)
memory usage: 22.0 KB


In [119]:
X_validate.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1409 entries, 3554 to 3108
Data columns (total 6 columns):
 #   Column         Non-Null Count  Dtype
---  ------         --------------  -----
 0   e_bill         1409 non-null   int64
 1   tenure         1409 non-null   int64
 2   Fiber          1409 non-null   uint8
 3   bank_transfer  1409 non-null   uint8
 4   e_check        1409 non-null   uint8
 5   add_ons        1409 non-null   int64
dtypes: int64(3), uint8(3)
memory usage: 48.2 KB


In [120]:
TN, FP, FN, TP = confusion_matrix(y_validate, rf_try_val).ravel()
#         print(f'\n{name}\n\n {rf}')
confusion(TN=TN, TP=TP, FN=FN, FP=FP)


    _______________________________________________________________________________________
    
    True Positive = 190 ---- False Positive = 114
    True Negative = 921 ---- False Negative = 184
    
    Out of 1409 predictions -- Correct predictions = 1111 (True Pos + True Neg) 
    
    REAL POSITIVE = (TP + FN) = 374 ---- PREDICTED POSITIVE = (TP + FP) = 304
    
    REAL NEGATIVE = (TN + FP) = 1035 ---- PREDICTED NEGATIVE = (TN + FN) = 1105
     
        Accuracy = 78.85% -->> Correct Predictions / Total Predictions
       Precision = 62.50% -->> True Positive / Predicted Positive
             NPV = 83.35% -->> True Negative / Predicted Negative
          Recall = 50.80% -->> True Positive / Real Positive
     Specificity = 88.99% -->> True Negative / Real Negative
        f1-score = 56.05% -->> Harmonic Mean of Precision and Recall
    _______________________________________________________________________________________
    


In [132]:
rf_try_2 = RandomForestClassifier(max_depth=8, max_samples=0.85, min_samples_leaf=3,
                       n_estimators=1000, random_state=123)

rf_try_2.fit(X_train,y_train)

pred_try_2_train = rf_try_2.predict(X_train)

pred_try_2_validate = rf_try_2.predict(X_validate)

TN, FP, FN, TP = confusion_matrix(y_train, pred_try_2_train).ravel()
print(f'\nTrain on \n\n{rf_try_2}')
confusion(TN=TN, TP=TP, FN=FN, FP=FP)

TNv, FPv, FNv, TPv = confusion_matrix(y_validate, pred_try_2_validate).ravel()
print(f'\nValidate on \n\n {rf_try_2}')
confusion(TN=TNv, TP=TPv, FN=FNv, FP=FPv)


Train on 

RandomForestClassifier(max_depth=8, max_samples=0.85, min_samples_leaf=3,
                       n_estimators=1000, random_state=123)

    _______________________________________________________________________________________
    
    True Positive = 594 ---- False Positive = 242
    True Negative = 2701 ---- False Negative = 520
    
    Out of 4057 predictions -- Correct predictions = 3295 (True Pos + True Neg) 
    
    REAL POSITIVE = (TP + FN) = 1114 ---- PREDICTED POSITIVE = (TP + FP) = 836
    
    REAL NEGATIVE = (TN + FP) = 2943 ---- PREDICTED NEGATIVE = (TN + FN) = 3221
     
        Accuracy = 81.22% -->> Correct Predictions / Total Predictions
       Precision = 71.05% -->> True Positive / Predicted Positive
             NPV = 83.86% -->> True Negative / Predicted Negative
          Recall = 53.32% -->> True Positive / Real Positive
     Specificity = 91.78% -->> True Negative / Real Negative
        f1-score = 60.92% -->> Harmonic Mean of Precision and Recall


In [139]:
rf_try_2 = RandomForestClassifier(max_depth=15, max_samples=0.5, min_samples_leaf=5,
                       n_estimators=1000, random_state=123)

rf_try_2.fit(X_train,y_train)

pred_try_2_train = rf_try_2.predict(X_train)

pred_try_2_validate = rf_try_2.predict(X_validate)

TN, FP, FN, TP = confusion_matrix(y_train, pred_try_2_train).ravel()
print(f'\nTrain on \n\n{rf_try_2}')
confusion(TN=TN, TP=TP, FN=FN, FP=FP)

TNv, FPv, FNv, TPv = confusion_matrix(y_validate, pred_try_2_validate).ravel()
print(f'\nValidate on ') # \n\n {rf_try_2}')
confusion(TN=TNv, TP=TPv, FN=FNv, FP=FPv)


Train on 

RandomForestClassifier(max_depth=15, max_samples=0.5, min_samples_leaf=5,
                       n_estimators=1000, random_state=123)

    _______________________________________________________________________________________
    
    True Positive = 592 ---- False Positive = 249
    True Negative = 2694 ---- False Negative = 522
    
    Out of 4057 predictions -- Correct predictions = 3286 (True Pos + True Neg) 
    
    REAL POSITIVE = (TP + FN) = 1114 ---- PREDICTED POSITIVE = (TP + FP) = 841
    
    REAL NEGATIVE = (TN + FP) = 2943 ---- PREDICTED NEGATIVE = (TN + FN) = 3216
     
        Accuracy = 81.00% -->> Correct Predictions / Total Predictions
       Precision = 70.39% -->> True Positive / Predicted Positive
             NPV = 83.77% -->> True Negative / Predicted Negative
          Recall = 53.14% -->> True Positive / Real Positive
     Specificity = 91.54% -->> True Negative / Real Negative
        f1-score = 60.56% -->> Harmonic Mean of Precision and Recall


In [145]:
logit = LogisticRegression(random_state=123, C=1.0)

logit.fit(X_train, y_train)

y_pred = logit.predict(X_train)
y_pred

y_pred1 = logit.predict(X_validate)
y_pred1

TN, FP, FN, TP = confusion_matrix(y_train, y_pred).ravel()
print(f'\nTrain on \n\n{rf_try_2}')
confusion(TN=TN, TP=TP, FN=FN, FP=FP)

TNv, FPv, FNv, TPv = confusion_matrix(y_validate, y_pred1).ravel()
print(f'\nValidate on ') # \n\n {rf_try_2}')
confusion(TN=TNv, TP=TPv, FN=FNv, FP=FPv)


Train on 

RandomForestClassifier(max_depth=15, max_samples=0.5, min_samples_leaf=5,
                       n_estimators=1000, random_state=123)

    _______________________________________________________________________________________
    
    True Positive = 534 ---- False Positive = 284
    True Negative = 2659 ---- False Negative = 580
    
    Out of 4057 predictions -- Correct predictions = 3193 (True Pos + True Neg) 
    
    REAL POSITIVE = (TP + FN) = 1114 ---- PREDICTED POSITIVE = (TP + FP) = 818
    
    REAL NEGATIVE = (TN + FP) = 2943 ---- PREDICTED NEGATIVE = (TN + FN) = 3239
     
        Accuracy = 78.70% -->> Correct Predictions / Total Predictions
       Precision = 65.28% -->> True Positive / Predicted Positive
             NPV = 82.09% -->> True Negative / Predicted Negative
          Recall = 47.94% -->> True Positive / Real Positive
     Specificity = 90.35% -->> True Negative / Real Negative
        f1-score = 55.28% -->> Harmonic Mean of Precision and Recall


In [146]:
logit = LogisticRegression(random_state=123, C=.5)

logit.fit(X_train, y_train)

y_pred = logit.predict(X_train)
y_pred

y_pred1 = logit.predict(X_validate)
y_pred1

TN, FP, FN, TP = confusion_matrix(y_train, y_pred).ravel()
print(f'\nTrain on \n\n{rf_try_2}')
confusion(TN=TN, TP=TP, FN=FN, FP=FP)

TNv, FPv, FNv, TPv = confusion_matrix(y_validate, y_pred1).ravel()
print(f'\nValidate on ') # \n\n {rf_try_2}')
confusion(TN=TNv, TP=TPv, FN=FNv, FP=FPv)


Train on 

RandomForestClassifier(max_depth=15, max_samples=0.5, min_samples_leaf=5,
                       n_estimators=1000, random_state=123)

    _______________________________________________________________________________________
    
    True Positive = 532 ---- False Positive = 284
    True Negative = 2659 ---- False Negative = 582
    
    Out of 4057 predictions -- Correct predictions = 3191 (True Pos + True Neg) 
    
    REAL POSITIVE = (TP + FN) = 1114 ---- PREDICTED POSITIVE = (TP + FP) = 816
    
    REAL NEGATIVE = (TN + FP) = 2943 ---- PREDICTED NEGATIVE = (TN + FN) = 3241
     
        Accuracy = 78.65% -->> Correct Predictions / Total Predictions
       Precision = 65.20% -->> True Positive / Predicted Positive
             NPV = 82.04% -->> True Negative / Predicted Negative
          Recall = 47.76% -->> True Positive / Real Positive
     Specificity = 90.35% -->> True Negative / Real Negative
        f1-score = 55.13% -->> Harmonic Mean of Precision and Recall


In [150]:
logit = LogisticRegression(random_state=123, C=1.0, penalty='none', solver='newton-cg')

logit.fit(X_train, y_train)

y_pred = logit.predict(X_train)
y_pred

y_pred1 = logit.predict(X_validate)
y_pred1

TN, FP, FN, TP = confusion_matrix(y_train, y_pred).ravel()
print(f'\nTrain on \n\n{rf_try_2}')
confusion(TN=TN, TP=TP, FN=FN, FP=FP)

TNv, FPv, FNv, TPv = confusion_matrix(y_validate, y_pred1).ravel()
print(f'\nValidate on ') # \n\n {rf_try_2}')
confusion(TN=TNv, TP=TPv, FN=FNv, FP=FPv)


Train on 

RandomForestClassifier(max_depth=15, max_samples=0.5, min_samples_leaf=5,
                       n_estimators=1000, random_state=123)

    _______________________________________________________________________________________
    
    True Positive = 534 ---- False Positive = 285
    True Negative = 2658 ---- False Negative = 580
    
    Out of 4057 predictions -- Correct predictions = 3192 (True Pos + True Neg) 
    
    REAL POSITIVE = (TP + FN) = 1114 ---- PREDICTED POSITIVE = (TP + FP) = 819
    
    REAL NEGATIVE = (TN + FP) = 2943 ---- PREDICTED NEGATIVE = (TN + FN) = 3238
     
        Accuracy = 78.68% -->> Correct Predictions / Total Predictions
       Precision = 65.20% -->> True Positive / Predicted Positive
             NPV = 82.09% -->> True Negative / Predicted Negative
          Recall = 47.94% -->> True Positive / Real Positive
     Specificity = 90.32% -->> True Negative / Real Negative
        f1-score = 55.25% -->> Harmonic Mean of Precision and Recall


In [152]:
logit = LogisticRegression(random_state=123, C=1.0, solver='liblinear', penalty='l1')

logit.fit(X_train, y_train)

y_pred = logit.predict(X_train)
y_pred

y_pred1 = logit.predict(X_validate)
y_pred1

TN, FP, FN, TP = confusion_matrix(y_train, y_pred).ravel()
print(f'\nTrain on \n\n{rf_try_2}')
confusion(TN=TN, TP=TP, FN=FN, FP=FP)

TNv, FPv, FNv, TPv = confusion_matrix(y_validate, y_pred1).ravel()
print(f'\nValidate on ') # \n\n {rf_try_2}')
confusion(TN=TNv, TP=TPv, FN=FNv, FP=FPv)


Train on 

RandomForestClassifier(max_depth=15, max_samples=0.5, min_samples_leaf=5,
                       n_estimators=1000, random_state=123)

    _______________________________________________________________________________________
    
    True Positive = 532 ---- False Positive = 282
    True Negative = 2661 ---- False Negative = 582
    
    Out of 4057 predictions -- Correct predictions = 3193 (True Pos + True Neg) 
    
    REAL POSITIVE = (TP + FN) = 1114 ---- PREDICTED POSITIVE = (TP + FP) = 814
    
    REAL NEGATIVE = (TN + FP) = 2943 ---- PREDICTED NEGATIVE = (TN + FN) = 3243
     
        Accuracy = 78.70% -->> Correct Predictions / Total Predictions
       Precision = 65.36% -->> True Positive / Predicted Positive
             NPV = 82.05% -->> True Negative / Predicted Negative
          Recall = 47.76% -->> True Positive / Real Positive
     Specificity = 90.42% -->> True Negative / Real Negative
        f1-score = 55.19% -->> Harmonic Mean of Precision and Recall


In [153]:
X_train

e_bill  tenure  Fiber  bank_transfer  e_check  add_ons
5911       1      26      0              1        0        1
2479       0      47      0              0        1        0
5889       1       1      0              0        0        1
6087       1      69      0              1        0        3
785        0      51      1              0        0        0
...      ...     ...    ...            ...      ...      ...
5269       1      60      1              0        1        1
6102       0       2      1              1        0        2
356        1      52      0              0        1        3
1756       1       4      1              0        1        0
2127       0      35      0              0        0        1

[4057 rows x 6 columns]

In [154]:
X_train_fiber = X_train[X_train.Fiber == 1]
X_train_fiber

e_bill  tenure  Fiber  bank_transfer  e_check  add_ons
785        0      51      1              0        0        0
3057       0      34      1              0        0        4
5757       1       4      1              0        1        0
3465       1       1      1              0        1        0
1539       0      64      1              0        1        2
...      ...     ...    ...            ...      ...      ...
4308       1      38      1              0        0        2
4998       1      41      1              0        1        2
5269       1      60      1              0        1        1
6102       0       2      1              1        0        2
1756       1       4      1              0        1        0

[1850 rows x 6 columns]

In [157]:
y_train_fiber = y_train
y_train_fiber

5911    0
2479    0
5889    1
6087    0
785     0
       ..
5269    0
6102    1
356     0
1756    1
2127    0
Name: churn1, Length: 4057, dtype: int64

In [160]:
y_train_fiber.drop(index=X_train[X_train.Fiber != 1].index.tolist(), inplace=True)

In [161]:
logit = LogisticRegression(random_state=123, C=1.0, solver='liblinear', penalty='l1')

logit.fit(X_train_fiber, y_train_fiber)

y_pred = logit.predict(X_train_fiber)
y_pred

y_pred1 = logit.predict(X_validate)
y_pred1

TN, FP, FN, TP = confusion_matrix(y_train_fiber, y_pred).ravel()
print(f'\nTrain on \n\n{rf_try_2}')
confusion(TN=TN, TP=TP, FN=FN, FP=FP)

TNv, FPv, FNv, TPv = confusion_matrix(y_validate, y_pred1).ravel()
print(f'\nValidate on ') # \n\n {rf_try_2}')
confusion(TN=TNv, TP=TPv, FN=FNv, FP=FPv)


Train on 

RandomForestClassifier(max_depth=15, max_samples=0.5, min_samples_leaf=5,
                       n_estimators=1000, random_state=123)

    _______________________________________________________________________________________
    
    True Positive = 493 ---- False Positive = 257
    True Negative = 821 ---- False Negative = 279
    
    Out of 1850 predictions -- Correct predictions = 1314 (True Pos + True Neg) 
    
    REAL POSITIVE = (TP + FN) = 772 ---- PREDICTED POSITIVE = (TP + FP) = 750
    
    REAL NEGATIVE = (TN + FP) = 1078 ---- PREDICTED NEGATIVE = (TN + FN) = 1100
     
        Accuracy = 71.03% -->> Correct Predictions / Total Predictions
       Precision = 65.73% -->> True Positive / Predicted Positive
             NPV = 74.64% -->> True Negative / Predicted Negative
          Recall = 63.86% -->> True Positive / Real Positive
     Specificity = 76.16% -->> True Negative / Real Negative
        f1-score = 64.78% -->> Harmonic Mean of Precision and Recall
  

# Bottom Cell
<br><br><br><br><br><br><br><br><br><br><br><br><br><br><br><br><br><br><br><br><br><br><br><br><br><br><br><br><br><br><br><br><br><br><br><br><br><br><br><br><br><br><br><br><br><br><br><br><br><br><br><br><br><br><br><br><br><br><br><br><br><br><br><br><br><br><br><br><br><br><br><br>